### PR TITLE
Refactor Purchase to use timeMachine

### LINKED ISSUE #5728: Change event used when renewing your purchase
**Describe the bug**
If you purchase 2 copies of the same key, some screens on etherscan consider this 2 tokens: https://etherscan.io/token/0x91002310cec24d818f42a169dd0d27ecbfa8bcd5?a=0xdb92c096bc5efa8adb48f05cd601dddb75228203#inventory

**To Reproduce**
1. purchase a key
2. purchase it again

**Expected behavior**
Per EIP-721 we should emit Transfer(from: 0) when minting.  But our use case is a bit unique.

I think we could address this by creating a new event type.  Emit Transfer only for the first purchase (when the tokenId is assigned).  Then when a token is renewed we can emit a new `Renewal` event instead.

### PR DESCRIPTION
**Pull request by nfurfaro**
# Description
This fixes the issues around events described in issue 5728, and simplifies the branching logic in `purchase` a fair bit.

# Issues

Fixes #5728 
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by nfurfaro** [comment]
Note: This needs to be rebased after PR #5859  is merged  to simplify review!

File: smart-contracts/test/Lock/purchaseFor.js [hunk 1]
```diff
@@ -16,252 +16,255 @@ contract('Lock / purchaseFor', accounts => {
...
   })
 
-  describe('when the contract has a public key release', () => {
-    it('should fail if the price is not enough', async () => {
-      await shouldFail(
-        locks.FIRST.purchase(0, accounts[0], web3.utils.padLeft(0, 40), [], {
-          value: Units.convert('0.0001', 'eth', 'wei'),
-        }),
-        'NOT_ENOUGH_FUNDS'
+  it('should fail if the price is not enough', async () => {
```
Comment (nfurfaro) [inline smart-contracts/test/Lock/purchaseFor.js]: Hmm. Removibg this ine "describe" statement is mak8ng it look kije Ive dobe waaay more in this fike than I actually did. I might restore it for the purpose of making reviewing easier.

Comment (nfurfaro) [reply smart-contracts/test/Lock/purchaseFor.js]: done.

**Comment by NickCuso** [comment]
I'm not sure we should emit 2 events for a standard purchase.

**Comment by NickCuso** [comment]
This is ~4-5k more expensive purchase.  What do we gain?

**Comment by NickCuso** [comment]
A more targeted bug fix may be worthwhile.  Let's aim to keep purchase clean

**Comment by nfurfaro** [comment]
Working on a cleaned up bug fix now.

**Comment by nfurfaro** [comment]
Closing this in favour of #5904.
cc @NickCuso

**Pull request closed without merging**